test-case/latency-metrics.sh

```diff
@@ -0,0 +1,287 @@
+elif [ -z "${pcm_p}" ] || [ -z "${pcm_c}" ]; then
+    skip_test "No playback or capture ALSA PCM is specified."
+else
+  JACKD_BACKEND_OPTIONS="-P ${pcm_p} -C ${pcm_c} ${JACKD_BACKEND_OPTIONS}"
+fi
+
+if [ "${set_loopback}" == 1 ]; then
+  port_playback="${set_loopback}"
+  port_capture="${set_loopback}"
+fi
+
+if [ -z "${port_playback}" ] || [ -z "${port_capture}" ]; then
+  skip_test "No playback or capture JACK port is specified."
+fi
+
+set_alsa()
+{
+  reset_sof_volume
+
+  # If MODEL is defined, set proper gain for the platform
+  if [ -z "$MODEL" ]; then
+    dlogw "NO MODEL is defined. Please define MODEL to run alsa_settings/MODEL.sh"
+  else
+    set_alsa_settings "$MODEL"
+  fi
+}
+
+check_jackd_configured()
+{
+  dlogi "Check JACK audio server can be started"
+  dlogc "jackd ${JACKD_OPTIONS} -d ${JACKD_BACKEND} ${JACKD_BACKEND_OPTIONS} -p ${jackd_period}"
+  # shellcheck disable=SC2086
+  JACK_NO_AUDIO_RESERVATION=1 timeout --kill-after ${RUN_PERIOD_MAX} ${run_verbose} ${JACKD_TIMEOUT} \
+      jackd ${JACKD_OPTIONS} -d ${JACKD_BACKEND} ${JACKD_BACKEND_OPTIONS} -p ${jackd_period} & jackdPID=$!
+
+  sleep ${WAIT_JACKD}
+
+  dlogc "jack_lsp"
+  jack_lsp -AclLpt
+
+  dlogi "Waiting jackd to stop without a client"
+  wait ${jackdPID}
+}
+
+collect_latency_data()
+{
+  dlogi "Start collecting latency data"
+  dlogc "jackd ${JACKD_OPTIONS} -d ${JACKD_BACKEND} ${JACKD_BACKEND_OPTIONS} -p ${jackd_period}"
+  # shellcheck disable=SC2086
+  JACK_NO_AUDIO_RESERVATION=1 timeout --kill-after ${RUN_PERIOD_MAX} ${run_verbose} ${JACKD_TIMEOUT} \
+      jackd ${JACKD_OPTIONS} -d ${JACKD_BACKEND} ${JACKD_BACKEND_OPTIONS} -p ${jackd_period} 2>&1 | \
+      tee >(AWKPATH="${TESTLIB}:${AWKPATH}" \
+          gawk -f "${TESTLIB}/jackd_events.awk" > "${EVENTS_JSON}") & jackdPID="$!"
+
+  sleep ${WAIT_JACKD}
+  dlogc "jack_iodelay"
+  # shellcheck disable=SC2086
+  timeout --kill-after ${RUN_PERIOD_MAX} ${run_verbose} ${IODELAY_TIMEOUT} \
+      stdbuf -oL -eL jack_iodelay | \
+      tee >(AWKPATH="${TESTLIB}:${AWKPATH}" \
+          gawk -f "${TESTLIB}/jack_iodelay_metrics.awk" > "${METRICS_JSON}") & iodelayPID="$!"
+
+  sleep ${WAIT_IODELAY}
+  if [ "${set_loopback}" == 1 ]; then
+      dlogi "jack_connect: [JACK loopback]<==>[jack_delay]"
+      jack_connect jack_delay:out jack_delay:in
+  else
+      dlogi "jack_connect: [${port_capture}]==>[jack_delay]==>[${port_playback}]"
+      jack_connect jack_delay:out "${port_playback}" && jack_connect jack_delay:in "${port_capture}"
+  fi
+
+  dlogi "Latency data collection"
+  wait ${jackdPID} ${iodelayPID}
+
+  dlogi "Latency data collection completed for period=${jackd_period} frames."
+}
+
+report_start()
+{
+  dlogi "Compose ${RESULT_JSON}"
+  echo -n "{\"options\":{$(options2json)}, " > "${RESULT_JSON}"
+  echo -n  "\"metrics\":[" >> "${RESULT_JSON}"
+}
+
+report_metric()
+{
+  echo -n  "${next_sep}{\"period_frames\":${jackd_period}, \"periods\":${jackd_nperiods}, \"rate\":${rate}, " >> "${RESULT_JSON}"
```

Comment:
This time it should be easy to avoid the backslashes, no "challenge"
https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo

Also, `%d` will fail if the input is not a number; a good thing.

```suggestion
  printf '%s{"%d":"%d", ...'  "${next_sep}" "period_frames" "${jackd_period}"  >> "${RESULT_JSON}"
```

OK, maybe just ONE backslash newline :-)

```suggestion
  printf '%s{"%d":"%d", ...'  \
     "${next_sep}" "period_frames" "${jackd_period}"  >> "${RESULT_JSON}"
```